domain/src/DomainForm.php

```diff
@@ -120,36 +155,40 @@ public function form(array $form, FormStateInterface $form_state) {
   /**
    * {@inheritdoc}
    */
-  public function validate(array $form, FormStateInterface $form_state) {
-    $entity = $this->buildEntity($form, $form_state);
-    $validator = \Drupal::service('domain.validator');
-    $errors = $validator->validate($entity);
+  public function validateForm(array &$form, FormStateInterface $form_state) {
+    /** @var \Drupal\domain\DomainInterface $entity */
+    $entity = $this->entity;
+    $hostname = $entity->getHostname();
+    $errors = $this->validator->validate($hostname);
     if (!empty($errors)) {
-      $form_state->setErrorByName('hostname', $errors);
+      // Render errors to display as message.
+      $message = [
+        '#theme' => 'item_list',
+        '#items' => $errors,
+      ];
+      $message = $this->renderer->renderPlain($message);
+      $form_state->setErrorByName('hostname', $message);
+    }
+    // Validate if the same hostname exists.
+    // Do not use domain loader because it may change hostname.
+    $existing = $this->storage->loadByProperties(['hostname' => $hostname]);
+    $existing = reset($existing);
+    if ($existing && $existing->id() != $entity->id()) {
```

Comment:
The intend here is to prevent to add another domain with same hostname, also it prevents edit to change hostname for existing

BTW I suppose entity level validation to properly handle REST operations on domain records